File: src/swingjs/JSAbstractDocument.java
package swingjs;

import java.util.HashMap;
import java.util.Map;

import swingjs.JSPlainDocument.JSElement;
import jsjava.awt.font.TextAttribute;
import jsjavax.swing.event.DocumentEvent;
import jsjavax.swing.event.DocumentListener;
import jsjavax.swing.event.EventListenerList;
import jsjavax.swing.event.UndoableEditEvent;
import jsjavax.swing.event.UndoableEditListener;
import jsjavax.swing.text.AttributeSet;
import jsjavax.swing.text.BadLocationException;
import jsjavax.swing.text.Document;
import jsjavax.swing.text.Element;
import jsjavax.swing.text.StyleConstants;
import jsjavax.swing.text.Utilities;
import jsjavax.swing.text.AbstractDocument.DefaultDocumentEvent;
import jsjavax.swing.undo.UndoableEdit;
import jssun.swing.SwingUtilities2;

public abstract class JSAbstractDocument  implements Document {
	protected Map<Object, Object> props;
	protected JSElement root;
	protected Map<Integer, JSPosition> positions;
	protected EventListenerList listenerList;
	private boolean notifyingListeners;

	public JSAbstractDocument() {
		props = new HashMap<Object, Object>();
	}
	
	@Override
	public Element[] getRootElements() {
		return new Element[] { root, null };
	}

	protected void fixPositions(int offset, int length, boolean isInsert) {
		if (positions == null || positions.isEmpty())
			return;
		if (isInsert) {
			for (Integer i : positions.keySet()) {
				int pos = i.intValue();
				if (pos > offset)
					positions.get(i).pos += length;
			}
			return;
		}
		for (Integer i : positions.keySet()) {
			int pos = i.intValue();
			if (pos <= offset)
				continue;
			if (pos >= offset + length)
				positions.get(i).pos -= length;
			else
				positions.get(i).pos = offset;
		}
	}

  /**
   * Performs the actual work of inserting the text; it is assumed the
   * caller has obtained a write lock before invoking this.
   */
	protected void handleInsertString(int offs, String str, AttributeSet a)
                       throws BadLocationException {
      if ((str == null) || (str.length() == 0)) {
          return;
      }
      // SwingJS n/a
      //UndoableEdit u = data.insertString(offs, str);
      JSDocumentEvent e =
          new JSDocumentEvent(this, offs, str.length(), DocumentEvent.EventType.INSERT);
//      if (u != null) {
//          e.addEdit(u);
//      }

//      // see if complex glyph layout support is needed
//      if( getProperty(I18NProperty).equals( Boolean.FALSE ) ) {
//          // if a default direction of right-to-left has been specified,
//          // we want complex layout even if the text is all left to right.
//          Object d = getProperty(TextAttribute.RUN_DIRECTION);
//          if ((d != null) && (d.equals(TextAttribute.RUN_DIRECTION_RTL))) {
//              putProperty( I18NProperty, Boolean.TRUE);
//          } else {
//              char[] chars = str.toCharArray();
//              if (SwingUtilities2.isComplexLayout(chars, 0, chars.length)) {
//                  putProperty( I18NProperty, Boolean.TRUE);
//              }
//          }
//      }
//
//      insertUpdate(e, a);
//      // Mark the edit as done.
//      e.end();
      fireInsertUpdate(e);
      // only fire undo if Content implementation supports it
      // undo for the composed text is not supported for now
//      if (u != null &&
//          (a == null || !a.isDefined(StyleConstants.ComposedTextAttribute))) {
//          fireUndoableEditUpdate(new UndoableEditEvent(this, e));
//      }
  }

	protected void checkLoc(int start, int end) throws BadLocationException {
		if (start < 0 || end > getLength())
			throw new BadLocationException("JSPlainDocument: out of range",
					(start < 0 ? start : end));
	}

  /**
   * Performs the actual work of the remove. It is assumed the caller
   * will have obtained a <code>writeLock</code> before invoking this.
   */
  protected void handleRemove(int offs, int len) throws BadLocationException {
      if (len > 0) {
          DocumentEvent chng =
                  new JSDocumentEvent(this, offs, len, DocumentEvent.EventType.REMOVE);

          //boolean isComposedTextElement = false;
          // Check whether the position of interest is the composed text
          //isComposedTextElement = Utilities.isComposedTextElement(this, offs);

          // SwingJS n/a
          // removeUpdate(chng);
          // SwingJS n/a 
          // UndoableEdit u = data.remove(offs, len);
          // if (u != null) {
          //     chng.addEdit(u);
          // }
          // SwingJS BIDI only
          // postRemoveUpdate(chng);
          // Mark the edit as done.
          ///chng.end();
          fireRemoveUpdate(chng);
          // only fire undo if Content implementation supports it
          // undo for the composed text is not supported for now
          // SwingJS n/a
          //if ((u != null) && !isComposedTextElement) {
          //    fireUndoableEditUpdate(new UndoableEditEvent(this, chng));
          //}
      }
  }

  /**
	 * Notifies all listeners that have registered interest for notification on
	 * this event type. The event instance is lazily created using the parameters
	 * passed into the fire method.
	 * 
	 * @param e
	 *          the event
	 * @see EventListenerList
	 */
	protected void fireInsertUpdate(DocumentEvent e) {
		if (listenerList == null)
			return;
		checkAlreadyNotifying();
		notifyingListeners = true;
		try {
			// Guaranteed to return a non-null array
			Object[] listeners = listenerList.getListenerList();
			// Process the listeners last to first, notifying
			// those that are interested in this event
			for (int i = listeners.length - 2; i >= 0; i -= 2) {
				if (listeners[i] == DocumentListener.class) {
					// Lazily create the event:
					// if (e == null)
					// e = new ListSelectionEvent(this, firstIndex, lastIndex);
					((DocumentListener) listeners[i + 1]).insertUpdate(e);
				}
			}
		} finally {
			notifyingListeners = false;
		}
	}

	/**
	 * SwingJS note: SwingJS does not implement redo/undo and compound edits.  
	 *   
	 * Notifies all listeners that have registered interest for notification on
	 * this event type. The event instance is lazily created using the parameters
	 * passed into the fire method.
	 * 
	 * @param e
	 *          the event
	 * @see EventListenerList
	 */
	protected void fireChangedUpdate(DocumentEvent e) {
		if (listenerList == null)
			return;
		checkAlreadyNotifying();
		notifyingListeners = true;
		try {
			// Guaranteed to return a non-null array
			Object[] listeners = listenerList.getListenerList();
			// Process the listeners last to first, notifying
			// those that are interested in this event
			for (int i = listeners.length - 2; i >= 0; i -= 2) {
				if (listeners[i] == DocumentListener.class) {
					// Lazily create the event:
					// if (e == null)
					// e = new ListSelectionEvent(this, firstIndex, lastIndex);
					((DocumentListener) listeners[i + 1]).changedUpdate(e);
				}
			}
		} finally {
			notifyingListeners = false;
		}
	}

	/**
	 * Notifies all listeners that have registered interest for notification on
	 * this event type. The event instance is lazily created using the parameters
	 * passed into the fire method.
	 * 
	 * @param e
	 *          the event
	 * @see EventListenerList
	 */
	protected void fireRemoveUpdate(DocumentEvent e) {
		if (listenerList == null)
			return;
		checkAlreadyNotifying();
		notifyingListeners = true;
		try {
			// Guaranteed to return a non-null array
			Object[] listeners = listenerList.getListenerList();
			// Process the listeners last to first, notifying
			// those that are interested in this event
			for (int i = listeners.length - 2; i >= 0; i -= 2) {
				if (listeners[i] == DocumentListener.class) {
					// Lazily create the event:
					// if (e == null)
					// e = new ListSelectionEvent(this, firstIndex, lastIndex);
					((DocumentListener) listeners[i + 1]).removeUpdate(e);
				}
			}
		} finally {
			notifyingListeners = false;
		}
	}
	
	private void checkAlreadyNotifying() {
		if (notifyingListeners)
			throw new IllegalStateException("One of the document listeners modifed the document. This is not allowed.");
	}

	@Override
	public void addDocumentListener(DocumentListener listener) {
		if (listenerList == null)
			listenerList = new EventListenerList();
		listenerList.add(DocumentListener.class, listener);
		// JFormattedTextField; DefaultCaret
	}

	@Override
	public void removeDocumentListener(DocumentListener listener) {
		if (listenerList != null)
			listenerList.remove(DocumentListener.class, listener);
	}

	@Override
	public void addUndoableEditListener(UndoableEditListener listener) {
		// SwingJS -- not necessary?
	}

	@Override
	public void removeUndoableEditListener(UndoableEditListener listener) {
		// SwingJS -- not necessary?
	}

	@Override
	public Object getProperty(Object key) {
		return props.get(key);
	}

	@Override
	public void putProperty(Object key, Object value) {
		props.put(key, value);
	}


}

File: site/swingjs/j2s/swingjs/JSAbstractDocument.js
Clazz.declarePackage ("swingjs");
Clazz.load (["javax.swing.text.DocumentFilter", "$.Element", "$.JSMinimalAbstractDocument"], "swingjs.JSAbstractDocument", ["java.lang.IllegalStateException", "java.util.HashMap", "JU.AU", "javax.swing.event.DocumentEvent", "$.DocumentListener", "$.EventListenerList", "javax.swing.text.BadLocationException", "swingjs.JSDocumentEvent"], function () {
c$ = Clazz.decorateAsClass (function () {
this.props = null;
this.root = null;
this.positions = null;
this.listenerList = null;
this.notifyingListeners = false;
this.filterBypass = null;
this.me = null;
this.sb = null;
this.tempChar = null;
this.filter = null;
if (!Clazz.isClassDefined ("swingjs.JSAbstractDocument.DefaultFilterBypass")) {
swingjs.JSAbstractDocument.$JSAbstractDocument$DefaultFilterBypass$ ();
}
if (!Clazz.isClassDefined ("swingjs.JSAbstractDocument.JSElement")) {
swingjs.JSAbstractDocument.$JSAbstractDocument$JSElement$ ();
}
Clazz.instantialize (this, arguments);
}, swingjs, "JSAbstractDocument", null, javax.swing.text.JSMinimalAbstractDocument);
Clazz.makeConstructor (c$, 
function () {
this.me = this;
this.props =  new java.util.HashMap ();
});
Clazz.overrideMethod (c$, "getRootElements", 
function () {
return [this.root, null];
});
Clazz.defineMethod (c$, "checkLoc", 
function (start, end) {
if (start < 0 || end > this.getLength ()) throw  new javax.swing.text.BadLocationException ("JSAbstractDocument: out of range", (start < 0 ? start : end));
}, "~N,~N");
Clazz.defineMethod (c$, "fixPositions", 
function (offset, length, isInsert) {
if (this.positions == null || this.positions.isEmpty ()) return;
if (isInsert) {
for (var i, $i = this.positions.keySet ().iterator (); $i.hasNext () && ((i = $i.next ()) || true);) {
var pos = i.intValue ();
if (pos > offset) this.positions.get (i).pos += length;
}
return;
}for (var i, $i = this.positions.keySet ().iterator (); $i.hasNext () && ((i = $i.next ()) || true);) {
var pos = i.intValue ();
if (pos <= offset) continue;
if (pos >= offset + length) this.positions.get (i).pos -= length;
 else this.positions.get (i).pos = offset;
}
}, "~N,~N,~B");
Clazz.defineMethod (c$, "getFilterBypass", 
 function () {
if (this.filterBypass == null) {
this.filterBypass = Clazz.innerTypeInstance (swingjs.JSAbstractDocument.DefaultFilterBypass, this, null);
}return this.filterBypass;
});
Clazz.overrideMethod (c$, "remove", 
function (offs, len) {
var filter = this.getDocumentFilter ();
if (filter == null) this.handleRemove (offs, len);
 else filter.remove (this.getFilterBypass (), offs, len);
}, "~N,~N");
Clazz.overrideMethod (c$, "insertString", 
function (offset, str, a) {
var filter = this.getDocumentFilter ();
if (filter == null) this.handleInsertString (offset, str, a);
 else filter.insertString (this.getFilterBypass (), offset, str, a);
}, "~N,~S,javax.swing.text.AttributeSet");
Clazz.overrideMethod (c$, "replace", 
function (offset, length, text, attrs) {
if (length == 0 && (text == null || text.length == 0)) return;
var filter = this.getDocumentFilter ();
if (filter != null) {
filter.replace (this.getFilterBypass (), offset, length, text, attrs);
} else {
if (length > 0) this.remove (offset, length);
if (text != null && text.length > 0) this.insertString (offset, text, attrs);
}}, "~N,~N,~S,javax.swing.text.AttributeSet");
Clazz.defineMethod (c$, "taint", 
 function () {
this.tempChar = null;
});
Clazz.defineMethod (c$, "setLines", 
 function () {
this.root = Clazz.innerTypeInstance (swingjs.JSAbstractDocument.JSElement, this, null);
var s = this.sb.toString ();
if (s.lastIndexOf ('\n') != s.length - 1) s += "\n";
var ilast = 0;
for (var i = 0; i < s.length; i++) {
if (s.charAt (i) != '\n') continue;
var e = Clazz.innerTypeInstance (swingjs.JSAbstractDocument.JSElement, this, null);
e.start = ilast;
e.end = i;
ilast = i + 1;
this.root.addChild (e);
}
});
Clazz.defineMethod (c$, "handleInsertString", 
function (offs, str, a) {
if ((str == null) || (str.length == 0)) {
return;
}this.checkLoc (offs, offs);
this.taint ();
this.sb.insert (offs, str);
this.fixPositions (offs, str.length, true);
if (str.indexOf ('\n') >= 0) this.setLines ();
var e =  new swingjs.JSDocumentEvent (this, offs, str.length, javax.swing.event.DocumentEvent.EventType.INSERT);
this.fireInsertUpdate (e);
}, "~N,~S,javax.swing.text.AttributeSet");
Clazz.defineMethod (c$, "handleRemove", 
function (offs, len) {
this.checkLoc (offs, offs + len);
this.taint ();
var str = this.sb.substring2 (offs, offs + len);
this.sb.replace (offs, offs + len, "");
this.fixPositions (offs, offs + len, false);
if (str.indexOf ('\n') >= 0) this.setLines ();
if (len > 0) {
var chng =  new swingjs.JSDocumentEvent (this, offs, len, javax.swing.event.DocumentEvent.EventType.REMOVE);
this.fireRemoveUpdate (chng);
}}, "~N,~N");
Clazz.defineMethod (c$, "fireInsertUpdate", 
function (e) {
if (this.listenerList == null) return;
this.checkAlreadyNotifying ();
this.notifyingListeners = true;
try {
var listeners = this.listenerList.getListenerList ();
for (var i = listeners.length - 2; i >= 0; i -= 2) {
if (listeners[i] === javax.swing.event.DocumentListener) {
(listeners[i + 1]).insertUpdate (e);
}}
} finally {
this.notifyingListeners = false;
}
}, "javax.swing.event.DocumentEvent");
Clazz.defineMethod (c$, "fireChangedUpdate", 
function (e) {
if (this.listenerList == null) return;
this.checkAlreadyNotifying ();
this.notifyingListeners = true;
try {
var listeners = this.listenerList.getListenerList ();
for (var i = listeners.length - 2; i >= 0; i -= 2) {
if (listeners[i] === javax.swing.event.DocumentListener) {
(listeners[i + 1]).changedUpdate (e);
}}
} finally {
this.notifyingListeners = false;
}
}, "javax.swing.event.DocumentEvent");
Clazz.defineMethod (c$, "fireRemoveUpdate", 
function (e) {
if (this.listenerList == null) return;
this.checkAlreadyNotifying ();
this.notifyingListeners = true;
try {
var listeners = this.listenerList.getListenerList ();
for (var i = listeners.length - 2; i >= 0; i -= 2) {
if (listeners[i] === javax.swing.event.DocumentListener) {
(listeners[i + 1]).removeUpdate (e);
}}
} finally {
this.notifyingListeners = false;
}
}, "javax.swing.event.DocumentEvent");
Clazz.defineMethod (c$, "checkAlreadyNotifying", 
 function () {
if (this.notifyingListeners) throw  new IllegalStateException ("One of the document listeners modifed the document. This is not allowed.");
});
Clazz.overrideMethod (c$, "addDocumentListener", 
function (listener) {
if (this.listenerList == null) this.listenerList =  new javax.swing.event.EventListenerList ();
this.listenerList.add (javax.swing.event.DocumentListener, listener);
}, "javax.swing.event.DocumentListener");
Clazz.overrideMethod (c$, "removeDocumentListener", 
function (listener) {
if (this.listenerList != null) this.listenerList.remove (javax.swing.event.DocumentListener, listener);
}, "javax.swing.event.DocumentListener");
Clazz.overrideMethod (c$, "addUndoableEditListener", 
function (listener) {
}, "javax.swing.event.UndoableEditListener");
Clazz.overrideMethod (c$, "removeUndoableEditListener", 
function (listener) {
}, "javax.swing.event.UndoableEditListener");
Clazz.overrideMethod (c$, "getProperty", 
function (key) {
return this.props.get (key);
}, "~O");
Clazz.overrideMethod (c$, "putProperty", 
function (key, value) {
this.props.put (key, value);
}, "~O,~O");
Clazz.overrideMethod (c$, "getAsynchronousLoadPriority", 
function () {
return -1;
});
Clazz.overrideMethod (c$, "setDocumentFilter", 
function (filter) {
this.filter = filter;
}, "javax.swing.text.DocumentFilter");
Clazz.defineMethod (c$, "getDocumentFilter", 
function () {
return this.filter;
});
c$.$JSAbstractDocument$DefaultFilterBypass$ = function () {
Clazz.pu$h(self.c$);
c$ = Clazz.decorateAsClass (function () {
Clazz.prepareCallback (this, arguments);
Clazz.instantialize (this, arguments);
}, swingjs.JSAbstractDocument, "DefaultFilterBypass", javax.swing.text.DocumentFilter.FilterBypass);
Clazz.overrideMethod (c$, "getDocument", 
function () {
return this.b$["swingjs.JSAbstractDocument"].me;
});
Clazz.overrideMethod (c$, "remove", 
function (a, b) {
this.b$["swingjs.JSAbstractDocument"].handleRemove (a, b);
}, "~N,~N");
Clazz.overrideMethod (c$, "insertString", 
function (a, b, c) {
this.b$["swingjs.JSAbstractDocument"].handleInsertString (a, b, c);
}, "~N,~S,javax.swing.text.AttributeSet");
Clazz.overrideMethod (c$, "replace", 
function (a, b, c, d) {
this.b$["swingjs.JSAbstractDocument"].handleRemove (a, b);
this.b$["swingjs.JSAbstractDocument"].handleInsertString (a, c, d);
}, "~N,~N,~S,javax.swing.text.AttributeSet");
c$ = Clazz.p0p ();
};
c$.$JSAbstractDocument$JSElement$ = function () {
Clazz.pu$h(self.c$);
c$ = Clazz.decorateAsClass (function () {
Clazz.prepareCallback (this, arguments);
this.parent = null;
this.attributeSet = null;
this.start = 0;
this.end = 0;
this.nchildren = 0;
this.children = null;
this.lastIndex = 0;
Clazz.instantialize (this, arguments);
}, swingjs.JSAbstractDocument, "JSElement", null, javax.swing.text.Element);
Clazz.makeConstructor (c$, 
function () {
this.children = null;
this.nchildren = 0;
this.lastIndex = -1;
});
Clazz.defineMethod (c$, "addChild", 
function (a) {
if (this.children == null) this.children =  new Array (10);
 else if (this.nchildren == this.children.length) this.children = JU.AU.doubleLength (this.children);
this.children[this.nchildren++] = a;
}, "swingjs.JSAbstractDocument.JSElement");
Clazz.overrideMethod (c$, "getDocument", 
function () {
return this.b$["swingjs.JSAbstractDocument"].me;
});
Clazz.overrideMethod (c$, "getParentElement", 
function () {
return this.parent;
});
Clazz.overrideMethod (c$, "getName", 
function () {
return this.getName ();
});
Clazz.overrideMethod (c$, "getAttributes", 
function () {
return this.attributeSet;
});
Clazz.defineMethod (c$, "getStartOffset", 
function () {
return this.start;
});
Clazz.defineMethod (c$, "getEndOffset", 
function () {
return this.end;
});
Clazz.overrideMethod (c$, "getElementIndex", 
function (a) {
var b;
var c = 0;
var d = this.nchildren - 1;
var e = 0;
var f = this.getStartOffset ();
var g;
if (this.nchildren == 0) {
return 0;
}if (a >= this.getEndOffset ()) {
return this.nchildren - 1;
}if ((this.lastIndex >= c) && (this.lastIndex <= d)) {
var h = this.children[this.lastIndex];
f = h.getStartOffset ();
g = h.getEndOffset ();
if ((a >= f) && (a < g)) {
return this.lastIndex;
}if (a < f) {
d = this.lastIndex;
} else {
c = this.lastIndex;
}}while (c <= d) {
e = c + (Clazz.doubleToInt ((d - c) / 2));
var h = this.children[e];
f = h.getStartOffset ();
g = h.getEndOffset ();
if ((a >= f) && (a < g)) {
b = e;
this.lastIndex = b;
return b;
} else if (a < f) {
d = e - 1;
} else {
c = e + 1;
}}
if (a < f) {
b = e;
} else {
b = e + 1;
}this.lastIndex = b;
return b;
}, "~N");
Clazz.overrideMethod (c$, "getElementCount", 
function () {
return this.nchildren;
});
Clazz.overrideMethod (c$, "getElement", 
function (a) {
return (a >= this.nchildren ? null : this.children[a]);
}, "~N");
Clazz.overrideMethod (c$, "isLeaf", 
function () {
return (this.parent != null);
});
c$ = Clazz.p0p ();
};
Clazz.defineStatics (c$,
"ParagraphElementName", "paragraph",
"ContentElementName", "content");
});
